lightningd/pay.c

```diff
@@ -314,9 +319,17 @@ local_routing_failure(const tal_t *ctx,
 	routing_failure->failcode = hout->failcode;
 	routing_failure->erring_node =
 	    tal_dup(routing_failure, struct node_id, &ld->id);
-	routing_failure->erring_channel = payment->route_channels[0];
-	routing_failure->channel_dir = node_id_idx(&ld->id,
-						   &payment->route_nodes[0]);
+
+	if (payment->route_nodes != NULL && payment->route_channels != NULL) {
+		routing_failure->erring_channel =
+		    tal_dup(routing_failure, struct short_channel_id,
+			    &payment->route_channels[0]);
+		routing_failure->channel_dir =
+		    node_id_idx(&ld->id, &payment->route_nodes[0]);
+	} else {
+		routing_failure->erring_channel = NULL;
```

Comment:
does the `channel_dir` not need to be set either?

Reply:
Leaving it unset on purpose so valgrind can tell us if we touch it inadvertently. We use the `erring_channel` as sentinel for the presence of both, and if we touch the direction we know it's an error (it's used in the JSON-RPC only anyway).